File: 백준/Gold/1300. K번째 수/K번째 수.py
import sys

n = int(sys.stdin.readline())
k = int(sys.stdin.readline())
start, end = 1, k
answer = 0

while start <= end:
  mid = (start + end) // 2
  cnt = 0

  for i in range(1, n+1):
    cnt += min(mid // i, n)

  if cnt >= k:
    end = mid -1
    answer = mid
  else:
    start = mid + 1

sys.stdout.write(str(answer) + "\n")

File: 백준/Gold/1300. K번째 수/K번째 수.cc
#include <iostream>
using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);

  int n, k;
  cin >> n >> k;
  int start = 0, end = k;
  int answer = 0;

  while (start <= end) {
    int mid = (start + end) / 2;
    int cnt = 0;

    for (int i = 1; i <= n; i++)
      cnt += min(mid / i, n);

    if (cnt >= k) {
      end = mid - 1;
      answer = mid;
    }
    else
      start = mid + 1;
  }

  cout << answer << "\n";

  return 0;
}